File: HelloTuiCpp/Classes/scene/Bag/Bagui.cpp
#include "Bagui.h"
#include "tagMap/tuiTag_bag.h"
#include "../../tui/TuiManager.h"

NS_BAG_BEGIN

void Bagui::onLoadScene()
{
	TuiManager::sharedManager()->parseScene(this, "panel_bag", PATH_BAG);

	CGridView* pGridView = (CGridView*)this->getControl(PANEL_BAG, GV_BAG);
	pGridView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_gvBag));
	pGridView->reloadData();

	CTableView* pTblView = (CTableView*)this->getControl(PANEL_BAG, TBL_BAG);
	pTblView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_tblBag));
	pTblView->reloadData();

	CPageView* pPageView = (CPageView*)this->getControl(PANEL_BAG, PV_BAG);
	pPageView->setDataSourceAdapter(this, ccw_datasource_adapter_selector(Bagui::event_adapt_pageBag));
	pPageView->reloadData();

	CButton* pBtnView = (CButton*)this->getControl(PANEL_BAG, BTN_BACK);
	pBtnView->setOnClickListener(this, ccw_click_selector(Bagui::event_btn_back));
}

CCObject* Bagui::event_adapt_gvBag(CCObject* pConvertCell, unsigned int uIdx)
{
	CGridViewCell* pCell = (CGridViewCell*)pConvertCell;
	CButton *btn = nullptr;
	if (!pCell)
	{
		pCell = new CGridViewCell();
		pCell->autorelease();

		TuiManager::sharedManager()->parseCell(pCell, "cell_grid", PATH_BAG);

		btn = (CButton*)pCell->getChildByTag(BTN_GRID);
		btn->setOnClickListener(this, ccw_click_selector(Bagui::event_btn_gv));
	}else{
		btn = (CButton*)pCell->getChildByTag(BTN_GRID);
	}
	btn->setUserTag(uIdx);
	return pCell;
}

CCObject* Bagui::event_adapt_tblBag(CCObject* pConvertCell, unsigned int uIdx)
{
	CTableViewCell* pCell = (CTableViewCell*)pConvertCell;
	if (!pCell)
	{
		pCell = new CTableViewCell();
		pCell->autorelease();

		TuiManager::sharedManager()->parseCell(pCell, "cell_grid", PATH_BAG);
	}else{

	}
	return pCell;
}

CCObject* Bagui::event_adapt_pageBag(CCObject* pConvertCell, unsigned int uIdx)
{
	CPageViewCell *pCell = (CPageViewCell*)pConvertCell;
	if (!pCell)
	{
		pCell = new CPageViewCell();
		pCell->autorelease();

		TuiManager::sharedManager()->parseCell(pCell, "cell_page", PATH_BAG);
	}else{

	}
	return pCell;
}

void Bagui::event_btn_back(CCObject* pSender)
{
	CCSceneManager::sharedManager()->replaceScene(
		CCSceneExTransitionFade::create(0.5f, LoadScene("Main::Mainui")));
}

void Bagui::event_btn_gv(CCObject* pSender)
{
	CButton *btn = (CButton*)pSender;
	int id = btn->getUserTag();

	CCLOG("id = %d", id);
}

/************************************************************************/
//	GET/SET/IS
/************************************************************************/
CCNode* Bagui::getPanel(int tagPanel)
{
	CCNode* pPanel = nullptr;
	switch (tagPanel)
	{
	case PANEL_BAG:
		pPanel = this->getChildByTag(PANEL_BAG);
		break;
	}
	return pPanel;
}

NS_BAG_END

File: HelloTuiLua/Resources/src/scene/bag/bagui.lua
require "extern"
require "tagMap/Tag_bagui"

Bagui = class("Bagui",function()
	return TuiBase:create()
end)

Bagui.__index = Bagui
local __instance = nil

function Bagui:create()
	local ret = Bagui.new()
	__instance = ret
	TuiManager:sharedManager():parseScene(ret,"panel_bag",PATH_BAGUI)
	ret:onLoadComplete()
	return ret
end

function Bagui:getControl(tagPanel,tagControl)
	local ret = nil
	ret = self:getPanel(tagPanel):getChildByTag(tagControl)
	return ret
end

function Bagui:getPanel(tagPanel)
	local ret = nil
	if tagPanel == Tag_bagui.PANEL_BAG then
		ret = self:getChildByTag(tagPanel)
	end
	return ret
end

---------------logic----------------------------

local function event_btn_back(p_sender)
	
	local scene = Welcomeui:create()
	CCDirector:sharedDirector():replaceScene(
		CCTransitionFade:create(0.5, scene, ccc3(0,255,255)))
end

local function event_btn_recombine(p_sender)

	local scene = Recombineui:create()
	CCDirector:sharedDirector():replaceScene(
		CCTransitionFade:create(0.5, scene, ccc3(0,255,255)))
end

local function event_btn_click(p_sender)
	print("tag : "..tolua.cast(p_sender,"CButton"):getUserTag())
end

local function event_adapt_gvbag(p_convertview, idx)
	local pCell = p_convertview
	local btn = nil
	if pCell == nil then
		pCell = CGridViewCell:new()
		pCell:autorelease()
		TuiManager:sharedManager():parseCell(pCell, "cell_grid", PATH_BAGUI)
		btn = tolua.cast(pCell:getChildByTag(Tag_bagui.BTN_GRID),"CButton")
		btn:setOnClickScriptHandler(event_btn_click)
	else
		btn = pCell:getChildByTag(Tag_bagui.BTN_GRID)
	end
	btn:setUserTag(idx)
	return pCell
end

local function event_adapt_tblbag(p_convertview, idx)
	local pCell = p_convertview
	if pCell == nil then
		pCell = CTableViewCell:new()
		pCell:autorelease()
		TuiManager:sharedManager():parseCell(pCell, "cell_grid", PATH_BAGUI)
	else

	end
	return pCell
end

local function event_adapt_pvbag(p_convertview, idx)
	local pCell = p_convertview
	if pCell == nil then
		pCell = CPageViewCell:new()
		pCell:autorelease()
		TuiManager:sharedManager():parseCell(pCell, "cell_page", PATH_BAGUI)
	else

	end
	return pCell
end

function Bagui:onLoadComplete()
	--加载完成
	local btnBack = tolua.cast(self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.BTN_BACK),"CButton")
	btnBack:setOnClickScriptHandler(event_btn_back)
	
	local btnRecombine = tolua.cast(self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.BTN_RECOMBINE),"CButton")
	btnRecombine:setOnClickScriptHandler(event_btn_recombine)
	
	local gvBag = tolua.cast(self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.GV_BAG),"CGridView")
	gvBag:setDataSourceAdapterScriptHandler(event_adapt_gvbag)
	gvBag:reloadData()
	
	local tblBag = tolua.cast(self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.TBL_BAG),"CTableView")
	tblBag:setDataSourceAdapterScriptHandler(event_adapt_tblbag)
	tblBag:reloadData()
	
	local pvBag = tolua.cast(self:getControl(Tag_bagui.PANEL_BAG,Tag_bagui.PV_BAG),"CPageView")
	pvBag:setDataSourceAdapterScriptHandler(event_adapt_pvbag)
	pvBag:reloadData()
	
end